### ISSUE TITLE
Performance Analysis on H20 32B Qwen Model

### ISSUE THREAD

**Issue by kzjeef** [kzjeef/dash-infer#2]
Nsys profiling:

# Profile command:

- bf16:
```
uv run nsys profile --delay 180 \
           --duration 30 --trace=cuda,nvtx   python3 benchmark_throughput.py \
           --model_path deepseek-ai/DeepSeek-R1-Distill-Qwen-32B --test_qps 1.0 --test_random_input --device_ids  0,1
```

- fp8:

```
uv run nsys profile --delay 280 \
                  --duration 30 --trace=cuda,nvtx       python3 benchmark_throughput.py --model_path deepseek-ai/DeepSeek-R1-Distill-Qwen-32B --test_qps 1.0 --test_random_input --device_ids  0,1 --quant_config '{"quant_method":"instant_quant","weight_format":"fp8_e4m3","compute_method":"activate_quant"}' --engine_max_batch 4
```

# Nsys files:

## FP16 profile file：
[qwen32b-bs32-decode-fp16-h20x2.nsys-rep.zip](https://github.com/user-attachments/files/21308956/qwen32b-bs32-decode-fp16-h20x2.nsys-rep.zip)

## 分析
取一次 decode：
这个 profile 受首包处理影响比较大。 decode 时间为500ms 左右（从两次 logits professor 中间取），没有找到不受首包影响的。 

## fp8 profile file：
https://drive.google.com/file/d/1D0L1yorl1vi40-iz-CZsMUBgdmxmImKi/view?usp=drive_link

## 分析：
这次使用 bs=4 来进行分析。
这个 profile 里面， decode 分为两种情况， 
一种不受 prefill 影响的， 一种是受 prefill 影响的， 
不影响的一次 decode 22ms

不受影响的 kernel 分布大约：
```
Time	Total Time	Instances	Avg	Med	Min	Max	StdDev	Name
23.5%	6.920 ms	384	18.020 μs	22.704 μs	7.104 μs	24.768 μs	7.559 μs	nvjet_qqtst_128x16_128x11_4x1_v_bz_algo2_TNT
18.8%	5.540 ms	128	43.282 μs	43.264 μs	42.817 μs	43.905 μs	196 ns	nvjet_qqtst_80x64_128x11_2x1_v_bz_algo2_TNN
12.0%	3.523 ms	128	27.527 μs	27.520 μs	27.200 μs	28.160 μs	166 ns	void span::detail::QKGemvKernel<(int)128, (int)128, (int)8, (int)16, (int)16, (int)8, (int)4, (int)1, (bool)0, cutlass::bfloat16_t, cutlass::bfloat16_t, span::qcache::QuantParam<(span::QuantMode)0, cutlass::bfloat16_t>, float, span::qcache::QCacheConfig<(span::QuantMode)0, cutlass::bfloat16_t>>(const T10 *, const T11 *const *, T10 *, const unsigned int *, const span::TileMapping *, T13, unsigned int, unsigned int, unsigned int)
8.2%	2.402 ms	258	9.311 μs	7.776 μs	7.008 μs	27.072 μs	2.988 μs	ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
7.7%	2.267 ms	128	17.709 μs	17.728 μs	16.896 μs	18.977 μs	523 ns	nvjet_qqtst_16x64_128x16_2x1_v_bz_algo2_bias_TNN
6.9%	2.045 ms	128	15.977 μs	15.936 μs	15.424 μs	17.504 μs	330 ns	void span::Kernel<span::detail::QKVGemmSm80Kernel<(int)128, (int)128, cutlass::bfloat16_t, cutlass::layout::RowMajor, cutlass::bfloat16_t, cutlass::layout::RowMajor, cutlass::bfloat16_t, cutlass::layout::RowMajor, float, cutlass::gemm::GemmShape<(int)16, (int)128, (int)32>, cutlass::gemm::GemmShape<(int)16, (int)128, (int)32>, cutlass::gemm::GemmShape<(int)16, (int)8, (int)16>, cutlass::epilogue::thread::LinearCombination<cutlass::bfloat16_t, (int)4, float, float, (cutlass::epilogue::thread::ScaleType::Kind)5, (cutlass::FloatRoundStyle)2, cutlass::bfloat16_t>, cutlass::gemm::threadblock::GemmBatchedIdentityThreadblockSwizzle, (int)4>>(T1::Params)
4.3%	1.257 ms	384	3.272 μs	2.336 μs	1.152 μs	6.400 μs	2.094 μs	void allspark::cuda::rotary_opt_kernel<__hie_buildin::bfloat16, (int)8, (int)128>(int, T1 *, const T1 *, const float *, const int *, int, int, int, int, int *, int, int, hie::Array<IntDivMod<unsigned int>, (int)2>, int *, int, int *)
4.2%	1.232 ms	640	1.924 μs	1.920 μs	1.632 μs	2.720 μs	177 ns	void allspark::cuda::per_tensor_symm_quantization_params_reduce_kernel<__hie_buildin::bfloat16, cutlass::float_e4m3_t, (int)256, (int)8>(const T1 *, float *, int)
3.0%	892.034 μs	640	1.393 μs	1.376 μs	1.280 μs	1.728 μs	67 ns	void allspark::cuda::per_tensor_symm_quantization_kernel<__hie_buildin::bfloat16, cutlass::float_e4m3_t, (int)128, (int)1, float>(const T1 *, const float *, T2 *, int)
2.5%	743.558 μs	258	2.882 μs	2.880 μs	2.624 μs	3.200 μs	96 ns	void allspark::cuda::layernorm_nobeta_kernel<(bool)0, (int)8, __hie_buildin::bfloat16>(T3 *, const T3 *, const T3 *, const T3 *, int, float)
```
API Summary:
```
Time	Total Time	Num Calls	Avg	Med	Min	Max	StdDev	Name
45.5%	14.566 ms	270	53.948 μs	44.994 μs	2.490 μs	227.734 μs	25.128 μs	cudaStreamSynchronize
30.1%	9.647 ms	3140	3.072 μs	2.930 μs	2.339 μs	56.659 μs	1.394 μs	cudaLaunchKernel
8.9%	2.858 ms	904	3.161 μs	2.930 μs	2.230 μs	60.389 μs	2.850 μs	cuLaunchKernelEx
6.7%	2.143 ms	910	2.354 μs	2.140 μs	1.760 μs	13.410 μs	757 ns	cudaMemsetAsync
5.6%	1.801 ms	568	3.170 μs	2.954 μs	1.960 μs	19.049 μs	1.366 μs	cudaMemcpyAsync
1.3%	419.007 μs	526	796 ns	740 ns	420 ns	10.179 μs	548 ns	cudaStreamWaitEvent
1.1%	349.870 μs	526	665 ns	620 ns	320 ns	7.670 μs	463 ns	cudaEventRecord
0.3%	101.707 μs	1284	79 ns	80 ns	49 ns	470 ns	36 ns	cuKernelGetAttribute
0.3%	91.308 μs	262	348 ns	355 ns	210 ns	770 ns	101 ns	cudaStreamGetCaptureInfo_v2_v11030
0.1%	40.750 μs	262	155 ns	130 ns	80 ns	560 ns	86 ns	cudaGetFuncBySymbol_v11000
```
这里面大约有 11ms 时间是在 launch kernel， 占掉了一半时间。 cuda graph 的重要性凸显。 

### Decode Stream和Prefill Stream 同时运行的时候的影响。

<img width="1532" height="459" alt="Image" src="https://github.com/user-attachments/assets/d357baa8-6626-418c-bf42-f2dd807b4caa" />
从这张图看， 在 prefill stream 上没有任务的时候， decode stream 的时间比较固定， kernel 也比较密集。 一次 decode 时间基本维持在 22ms 左右。
但是在 prefill stream 上有任务的时候， decode stream 上的 kernel间隔变得非常大。 
从logits 这个kernel 看 step 的时间也变成了 300+ms：

<img width="1100" height="488" alt="Image" src="https://github.com/user-attachments/assets/fa227df6-1c7b-42f7-bbc1-7eaf268af552" />

看前 25 个kernel 时间统计：
```
Time	Total Time	Instances	Avg	Med	Min	Max	StdDev	Name
34.1%	279.899 ms	68	4.116 ms	4.102 ms	4.055 ms	4.452 ms	89.217 μs	nvjet_qqtst_80x128_128x8_1x2_h_bz_algo2_TNN
17.3%	141.744 ms	34	4.169 ms	4.166 ms	4.081 ms	4.254 ms	50.357 μs	void flash_fwd_kernel<Flash_fwd_kernel_traits<(int)128, (int)128, (int)64, (int)4, (bool)0, (bool)0, cutlass::bfloat16_t, Flash_kernel_traits<(int)128, (int)128, (int)64, (int)4, cutlass::bfloat16_t>>, (bool)0, (bool)1, (bool)0, (bool)0, (bool)0, (bool)1, (bool)0, (bool)0>(Flash_fwd_params)
17.1%	140.587 ms	34	4.135 ms	4.112 ms	4.078 ms	4.482 ms	101.007 μs	nvjet_qqtst_128x128_128x6_1x2_h_bz_coopA_algo2_TNT
11.7%	96.131 ms	326	294.879 μs	40.656 μs	7.840 μs	9.449 ms	814.397 μs	ncclDevKernel_AllReduce_Sum_bf16_RING_LL(ncclDevKernelArgsStorage<(unsigned long)4096>)
4.7%	38.917 ms	36	1.081 ms	1.071 ms	1.064 ms	1.200 ms	33.645 μs	nvjet_qqtst_64x256_128x5_2x1_v_bz_coopB_algo2_bias_TNT
3.3%	26.924 ms	34	791.869 μs	769.859 μs	763.459 μs	860.708 μs	32.614 μs	nvjet_qqtst_64x256_128x5_2x1_v_bz_coopB_algo2_TNT
1.5%	12.410 ms	326	38.069 μs	3.344 μs	2.688 μs	186.497 μs	67.334 μs	void allspark::cuda::layernorm_nobeta_kernel<(bool)0, (int)8, __hie_buildin::bfloat16>(T3 *, const T3 *, const T3 *, const T3 *, int, float)
1.5%	12.116 ms	810	14.958 μs	1.920 μs	1.312 μs	160.928 μs	29.851 μs	void allspark::cuda::per_tensor_symm_quantization_kernel<__hie_buildin::bfloat16, cutlass::float_e4m3_t, (int)128, (int)1, float>(const T1 *, const float *, T2 *, int)
1.4%	11.836 ms	384	30.824 μs	30.288 μs	7.136 μs	155.552 μs	18.723 μs	nvjet_qqtst_128x16_128x11_4x1_v_bz_algo2_TNT
1.3%	10.637 ms	128	83.103 μs	83.520 μs	42.913 μs	288.545 μs	52.514 μs	nvjet_qqtst_80x64_128x11_2x1_v_bz_algo2_TNN
0.9%	7.584 ms	810	9.363 μs	2.528 μs	1.632 μs	67.712 μs	15.155 μs	void allspark::cuda::per_tensor_symm_quantization_params_reduce_kernel<__hie_buildin::bfloat16, cutlass::float_e4m3_t, (int)256, (int)8>(const T1 *, float *, int)
0.9%	7.137 ms	162	44.056 μs	2.928 μs	1.984 μs	219.137 μs	80.042 μs	void allspark::cuda::elementwise::ApplyGeneric<(int)8, (bool)0, allspark::cuda::elementwise::SimpleFactory<allspark::cuda::SiluFunctor<__hie_buildin::bfloat16>>, __hie_buildin::bfloat16, __hie_buildin::bfloat16>(T3, long, allspark::cuda::elementwise::GetPackType<T4, T1>::type *, const allspark::cuda::elementwise::GetPackType<T5, T1>::type *..., long, T4 *, const T5 *...)
0.9%	6.998 ms	162	43.194 μs	2.368 μs	1.792 μs	221.345 μs	79.524 μs	void allspark::cuda::elementwise::ApplyGeneric<(int)8, (bool)0, allspark::cuda::elementwise::SimpleFactory<allspark::cuda::MulFunctor<__hie_buildin::bfloat16>>, __hie_buildin::bfloat16, __hie_buildin::bfloat16, __hie_buildin::bfloat16>(T3, long, allspark::cuda::elementwise::GetPackType<T4, T1>::type *, const allspark::cuda::elementwise::GetPackType<T5, T1>::type *..., long, T4 *, const T5 *...)
0.8%	6.358 ms	486	13.082 μs	4.304 μs	1.280 μs	119.265 μs	24.094 μs	void allspark::cuda::rotary_opt_kernel<__hie_buildin::bfloat16, (int)8, (int)128>(int, T1 *, const T1 *, const float *, const int *, int, int, int, int, int *, int, int, hie::Array<IntDivMod<unsigned int>, (int)2>, int *, int, int *)
0.7%	5.422 ms	324	16.733 μs	2.400 μs	1.536 μs	95.328 μs	28.036 μs	void allspark::cuda::elementwise::ApplyGeneric<(int)8, (bool)0, allspark::cuda::elementwise::SimpleFactory<allspark::cuda::AddFunctor<__hie_buildin::bfloat16>>, __hie_buildin::bfloat16, __hie_buildin::bfloat16, __hie_buildin::bfloat16>(T3, long, allspark::cuda::elementwise::GetPackType<T4, T1>::type *, const allspark::cuda::elementwise::GetPackType<T5, T1>::type *..., long, T4 *, const T5 *...)
0.5%	4.221 ms	128	32.977 μs	28.512 μs	21.728 μs	108.320 μs	17.910 μs	void span::detail::QKGemvKernel<(int)128, (int)128, (int)8, (int)16, (int)16, (int)8, (int)4, (int)1, (bool)0, cutlass::bfloat16_t, cutlass::bfloat16_t, span::qcache::QuantParam<(span::QuantMode)0, cutlass::bfloat16_t>, float, span::qcache::QCacheConfig<(span::QuantMode)0, cutlass::bfloat16_t>>(const T10 *, const T11 *const *, T10 *, const unsigned int *, const span::TileMapping *, T13, unsigned int, unsigned int, unsigned int)
0.5%	4.054 ms	128	31.673 μs	27.776 μs	21.984 μs	170.080 μs	18.482 μs	nvjet_qqtst_16x64_128x16_2x1_v_bz_algo2_bias_TNN
0.3%	2.133 ms	128	16.661 μs	16.000 μs	13.056 μs	30.112 μs	3.036 μs	void span::Kernel<span::detail::QKVGemmSm80Kernel<(int)128, (int)128, cutlass::bfloat16_t, cutlass::layout::RowMajor, cutlass::bfloat16_t, cutlass::layout::RowMajor, cutlass::bfloat16_t, cutlass::layout::RowMajor, float, cutlass::gemm::GemmShape<(int)16, (int)128, (int)32>, cutlass::gemm::GemmShape<(int)16, (int)128, (int)32>, cutlass::gemm::GemmShape<(int)16, (int)8, (int)16>, cutlass::epilogue::thread::LinearCombination<cutlass::bfloat16_t, (int)4, float, float, (cutlass::epilogue::thread::ScaleType::Kind)5, (cutlass::FloatRoundStyle)2, cutlass::bfloat16_t>, cutlass::gemm::threadblock::GemmBatchedIdentityThreadblockSwizzle, (int)4>>(T1::Params)
0.2%	1.643 ms	68	24.167 μs	23.488 μs	22.336 μs	50.944 μs	3.638 μs	void allspark::cuda::cache::QuantSpanCopyKernel<(int)32, (span::QuantMode)0, (int)128, (int)128, __hie_buildin::bfloat16>(allspark::cuda::qcache::QCacheConfig<T2, T5>::QuantT *const *, const T5 *, unsigned int)
0.2%	1.321 ms	34	38.857 μs	37.392 μs	36.608 μs	72.384 μs	6.294 μs	void allspark::cuda::update_kv_kernel<__hie_buildin::bfloat16>(T1 *, T1 *, const T1 *, const T1 *, int, int, int, int, int, int)
0.1%	591.236 μs	2	295.618 μs	295.618 μs	287.010 μs	304.226 μs	12.173 μs	nvjet_tst_512x8_64x3_2x1_v_bz_NNT
0.1%	557.762 μs	128	4.357 μs	3.840 μs	2.880 μs	7.360 μs	1.349 μs	void span::InplaceSoftmaxKernel<(int)512, (int)1, (int)16, cutlass::bfloat16_t, float>(T4 *, const unsigned int *, span::U32DivMod, unsigned int, unsigned int)
0.0%	406.979 μs	128	3.179 μs	2.576 μs	1.984 μs	14.048 μs	1.642 μs	void span::detail::QKVReduceKernel<(int)128, (int)16, (int)16, (int)8, (int)4, (int)128, cutlass::bfloat16_t, float>(const T7 *, T7 *, const unsigned int *, unsigned int, span::U32DivMod)
0.0%	319.682 μs	128	2.497 μs	2.304 μs	1.888 μs	6.016 μs	624 ns	void allspark::cuda::cache::QuantCacheAppendKernel<(int)32, (int)4, (int)128, (int)128, (span::QuantMode)0, __hie_buildin::bfloat16>(allspark::cuda::qcache::QCacheConfig<T5, T6>::QuantT *const *, allspark::cuda::qcache::QCacheConfig<T5, T6>::QuantT *const *, T6 *, const T6 *, const unsigned int *, unsigned int, unsigned int)
```
如果看 decode 里面的几个 kernel 的运行时间，本身的时间几乎没有变化，比如void span::detail::QKGemvKernel 这个的时间还是28.5us， 128 次， 所以只是间隔变大了。 

然后看API 统计
```
Time	Total Time	Num Calls	Avg	Med	Min	Max	StdDev	Name
96.5%	1.377 s	342	4.027 ms	1.150 ms	3.170 μs	14.262 ms	4.730 ms	cudaStreamSynchronize
1.7%	24.322 ms	3922	6.201 μs	4.410 μs	2.390 μs	305.653 μs	14.472 μs	cudaLaunchKernel
0.8%	11.082 ms	634	17.479 μs	4.420 μs	1.920 μs	3.411 ms	188.775 μs	cudaMemcpyAsync
0.5%	6.813 ms	1250	5.450 μs	3.620 μs	1.900 μs	504.108 μs	22.527 μs	cudaMemsetAsync
0.4%	6.283 ms	1142	5.501 μs	4.500 μs	2.640 μs	207.135 μs	9.318 μs	cuLaunchKernelEx
0.0%	630.141 μs	662	951 ns	640 ns	300 ns	102.297 μs	4.069 μs	cudaEventRecord
0.0%	615.106 μs	662	929 ns	800 ns	380 ns	9.629 μs	643 ns	cudaStreamWaitEvent
0.0%	205.930 μs	330	624 ns	480 ns	220 ns	17.890 μs	1.052 μs	cudaStreamGetCaptureInfo_v2_v11030
0.0%	199.105 μs	1624	122 ns	80 ns	49 ns	5.690 μs	167 ns	cuKernelGetAttribute
0.0%	104.946 μs	330	318 ns	300 ns	80 ns	1.030 μs	176 ns	cudaGetFuncBySymbol_v11000
```
这里 kernel luanch 占比相对就很少了。
